src/Npgsql/TypeHandlers/ArrayHandler.cs

```diff
@@ -209,130 +148,177 @@ public override ValueTask<Array> Read(NpgsqlReadBuffer buf, int len, bool async,
             }
         }
 
+        protected async ValueTask<List<T>> ReadList<T>(NpgsqlReadBuffer buf, bool async)
+        {
+            await buf.Ensure(12, async);
+            var dimensions = buf.ReadInt32();
+
+            if (dimensions == 0)
+                return new List<T>();
+            if (dimensions > 1)
+                throw new NotSupportedException($"Can't read multidimensional array as List<{typeof(T).Name}>");
+
+            buf.ReadInt32();  // Has nulls. Not populated by PG?
+            buf.ReadUInt32(); // Element OID. Ignored.
+
+            await buf.Ensure(8, async);
+            var length = buf.ReadInt32();
+            buf.ReadInt32(); // We don't care about the lower bounds
+
+            var list = new List<T>(length);
+            for (var i = 0; i < length; i++)
+                list.Add(await ElementHandler.ReadWithLength<T>(buf, async));
+            return list;
+        }
+
         #endregion
 
         #region Write
 
-        public override int ValidateAndGetLength<T2>(T2 value, ref NpgsqlLengthCache lengthCache, NpgsqlParameter parameter)
-            => ValidateAndGetLength<TElement>(value, ref lengthCache, parameter);
+        static Exception MixedTypesOrJaggedArray(Exception innerException)
+            => new Exception("While trying to write an array, one of its elements failed validation. " +
+                             "You may be trying to mix types in a non-generic IList, or to write a jagged array.", innerException);
 
-        protected internal override int ValidateObjectAndGetLength(object value, ref NpgsqlLengthCache lengthCache, NpgsqlParameter parameter = null)
-            => ValidateAndGetLength<TElement>(value, ref lengthCache, parameter);
+        static Exception CantWriteType(Type type)
+            => new InvalidCastException($"Can't write type {type} as an array of {typeof(TElement)}");
 
         // We're required to override this but it will never be called
         public override int ValidateAndGetLength(Array value, ref NpgsqlLengthCache lengthCache, NpgsqlParameter parameter)
-            => throw new NotSupportedException();
-
-        public int ValidateAndGetLength<TElement2>(object value, ref NpgsqlLengthCache lengthCache, NpgsqlParameter parameter = null)
         {
             if (lengthCache == null)
                 lengthCache = new NpgsqlLengthCache(1);
             if (lengthCache.IsPopulated)
                 return lengthCache.Get();
+            return ValidateAndGetLengthNonGeneric(value, ref lengthCache);
```

Comment:
I think this method still doesn't ever get called (as the comment above says), because we override `ValidateAndGetLength<TAny>` below, right? If so, better to throw `NotSupportedException` as before?

Reply:
It's a public API now (a public member in a public class since #1885). This means that someone could use `ArrayHandler` for its own purposes like a plugin, and throwing `NotSupportedException` isn't a good idea.

Reply:
Hmm... I'm not too sure about this. Since ArrayHandler needs to support both single-dimensional arrays, multi-dimensional arrays, and now also lists, it overrides `ValidateAndGetLength<TAny>` rather than the more standard `ValidateAndGetLength(TDefault value...)`. It's a pretty special case in the type handlers, I'm not really sure if there's another one.

Now, I can't really imagine a case  where someone would actually call `ValidateAndGetLength(TDefault value...)` directly - the whole point of this method is to be called from the default `ValidateAndGetLength<TAny>`, which is overridden in the case of ArrayHandler and implemented completely differently... In other words, yes, it's public, but it's meant to be called from inside Npgsql and not from the outside, from a plugin.

To summarize, I think any case in which `ValidateAndGetLength(TDefault value...)` is called from the outside is a bug. If that's the case, then it should probably throw (and even be sealed).

Reply:
I think you're right...! ArrayHandler currently seems to override all the functionality provided by `NpgsqlTypeHandler<T>`, so it would be cleaner for it to just override `NpgsqlTypeHandler` directly.

Well done... do you want to do this as a part of this PR?

Reply:
Let's do it in an additional PR, but to finish this I would like to use `NotSupprotedException` as you suggested.

Reply:
No problem (I suspect this will actually not be very complicated to do). So I'll let you change to `NotSupportedException` (and also get rid of the local Impl methods) and then you can merge.